File: src/main/java/org/goafabric/personservice/logic/PersonLogic.java
package org.goafabric.personservice.logic;

import org.goafabric.personservice.adapter.CalleeServiceAdapter;
import org.goafabric.personservice.controller.dto.Person;
import org.goafabric.personservice.controller.dto.PersonSearch;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
//@CircuitBreaker(name = "calleeservice")
public class PersonLogic {


    private final CalleeServiceAdapter calleeServiceAdapter;

    public PersonLogic(CalleeServiceAdapter calleeServiceAdapter) {

        this.calleeServiceAdapter = calleeServiceAdapter;
    }

    public Person getById(String id) {
        return null;
    }

    public List<Person> find(PersonSearch personSearch, Integer page, Integer size) {
      return null;
    }

    public List<Person> findByStreet(String street, Integer page, Integer size) {
       return null;
    }

    public Person save(Person person) {
       return null;
    }

    public Person sayMyName(String name) {
        return new Person(null, null,
                calleeServiceAdapter.sayMyName(name).message(), "", null);
    }

}

File: src/main/kotlin/org/goafabric/personservice/logic/PersonLogic.kt
package org.goafabric.personservice.logic

import org.goafabric.personservice.adapter.CalleeServiceAdapter
import org.goafabric.personservice.controller.dto.Person
import org.goafabric.personservice.controller.dto.PersonSearch
import org.goafabric.personservice.persistence.PersonRepository
import org.springframework.data.domain.Example
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
@Transactional
class PersonLogic(
    private val personMapper: PersonMapper,
    private val personRepository: PersonRepository,
    private val calleeServiceAdapter: CalleeServiceAdapter
) {
    fun getById(id: String): Person {
        return personMapper.map(
            personRepository.findById(id).orElseThrow()
        )
    }

    fun find(personSearch: PersonSearch, page: Int, size: Int): List<Person> {
        return personMapper.map(
            personRepository.findAll(
                Example.of(personMapper.map(personSearch)),
                PageRequest.of(page, size)
            )
        )
    }

    fun findByStreet(street: String, page: Int, size: Int): List<Person> {
        return personMapper.map(
            personRepository.findByAddressStreetContainsIgnoreCase(street, PageRequest.of(page, size))
        )
    }

    fun save(person : Person): Person {
        return personMapper.map(personRepository.save(
            personMapper.map(person))
        )
    }

    fun sayMyName (name : String) : Person {
        return Person(firstName = calleeServiceAdapter.sayMyName(name)!!.message, lastName = "", address = emptyList())
    }

}
